### PR TITLE
Standalone program to parse Debezium Transaction messages

### PR DESCRIPTION
**Pull request by cirego**
This adds a new binary to the `demo/debezium` work that will read transaction messages from the debezium transaction topic and convert them into structured objects.

The validation work is still TBD. I'm putting this PR up because I felt like it got lost in the weeds of codifying how to parse Debezium messages / our Avro parsing. I've also found a few opportunities for improvement in materialized debezium parsing and will open a PR for that as well.

Long term, I think the program is useful to keep around because it can provide a reference for how to write a Debezium consumer that can be simpler than the materialized version.

### CODE REVIEW

**Comment by cuongdo** [comment]
@ruchirK could you have a look?

**Comment by ruchirK** [comment]
So - without having read the code I'm not entirely sure what the long term intention is with this standalone program.

what is it going to validate? (my guess is that it will validate whether debezium transaction messages come in order or not but I am not sure)

how do we expect to use it? (again, based on my guess this seems more like a program that validates materialize's inputs rather than its output so I'm not sure if this is being written as a test to help develop a feature or if we intend to use this in say CI or something like that)

**Comment by cirego** [comment]
> So - without having read the code I'm not entirely sure what the long term intention is with this standalone program.
> 
> what is it going to validate? (my guess is that it will validate whether debezium transaction messages come in order or not but I am not sure)
> 
> how do we expect to use it? (again, based on my guess this seems more like a program that validates materialize's inputs rather than its output so I'm not sure if this is being written as a test to help develop a feature or if we intend to use this in say CI or something like that)

Hi @ruchirK, my plan for this standalone program is two-fold:

1. Provide a simplified, reference implementation for how to write a Debezium consumer. I imagine that the Materialize implementation will become complicated over time and having a simpler implementation can help when trying to determine if a bug is in Materialize or Debezium.
2. Add validation checks to this implementation for things like "if we have read transactions [A, B, C, D] and the next transaction we see is B, then we expect to see [C, D] following" (assuming this is a proper thing to assert on!).

**Comment by ruchirK** [comment]
> Add validation checks to this implementation for things like "if we have read transactions [A, B, C, D] and the next transaction we see is B, then we expect to see [C, D] following" (assuming this is a proper thing to assert on!).

Can you say more about how this would work? as in - would the validation know the set of transactions in advance of validating the topic or would it be more like an online analysis ("as I scan through the txn topic, transaction ids must be increasing and noninterleaved)

I'm still not sure when we would run this would it be:

a) manually when testing with Debezium data to see if its correct
b) as part of a pre-merge / nightly CI run?
c) alongside the Materialize binary?

**Comment by cirego** [comment]
> > Add validation checks to this implementation for things like "if we have read transactions [A, B, C, D] and the next transaction we see is B, then we expect to see [C, D] following" (assuming this is a proper thing to assert on!).
> 
> Can you say more about how this would work? as in - would the validation know the set of transactions in advance of validating the topic or would it be more like an online analysis ("as I scan through the txn topic, transaction ids must be increasing and noninterleaved)
> 
> I'm still not sure when we would run this would it be:
> 
> a) manually when testing with Debezium data to see if its correct
> b) as part of a pre-merge / nightly CI run?
> c) alongside the Materialize binary?

My plan is to have this scan a set of topics that already have messages, with the transaction being the first such topic. As I mentioned above, I opened this PR because I felt that the decoding logic was getting sufficiently complicated that I wanted feedback on how it looked.

Today, I'm running this manually. Longer term, the plan would be to run it was part of our normal CI, as well as when testing new versions of Debezium. I don't have plans to deploy this anywhere as a long-lived process.

File: demo/debezium/validate-topics/src/bin/validate-debezium-topics.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+fn parse_transaction_id(fields: &HashMap<String, Value>) -> Result<String, anyhow::Error> {
+    match fields.get("id") {
+        Some(transaction) => match transaction {
+            Value::String(t) => Ok(t.to_string()),
+            _ => Err(anyhow!(
+                "Expected transaction to be a String but got: {:?}",
+                transaction
+            )),
+        },
+        None => Err(anyhow!("Expected to find a id field!")),
```
Comment (ruchirK) [inline demo/debezium/validate-topics/src/bin/validate-debezium-topics.rs:211]: `bail!` is the same as `Err(anyhow!` iiuc

**Review by ruchirK — APPROVED** [review]
I've lightly skimmed this and it looks reasonable so far. My understanding from our call a few days ago is that this isn't meant to be merged soon, and will instead be used to help test out our fixes for #6296 / help validate invariants in incoming data. is that right?

**Comment by cirego** [comment]
I've added two invariant checks, one for event_count and the other for the collections counts.

**Pull request merged**